### PR TITLE
Check for store api rest route in parameter

### PR DESCRIPTION
**Pull request by bborman22**
#### Changes proposed in this Pull Request

This PR addresses an issue we found where WooPay (p1698264612175849-slack-C022WMN88KG) would not work for sites that were using the Plain permalinks setting. The issue presented itself as the shopper getting an empty cart warning once they were on WooPay. What was happening is that the `WCPay\Platform_Checkout\SessionHandler` wasn't being used to set the session, so the Cart-Token wasn't being used to initialize the session. The WooPay SessionHandler wasn't being used because it is only set under certain conditions, one of which was the `is_store_api_request` check. This check was only looking for routes in the URI, while when using Plain permalinks the route is in a parameter like so: `https://domain.com/index.php?rest_route=/wc/store/v1/cart`

#### Testing instructions

The test matrix for this will involve testing with: [ first party auth on, first party auth off] and [ post name permalinks, plain permalinks].

For each scenario we should confirm that we are able to perform a successful transaction.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7558` or branch name `fix/woopay-session-handler-permalinks` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/woopay-session-handler-permalinks) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 7b7744ec22f932bc2660de0c90dfece5157567c3
- Build time: 2023-11-02 13:46:37 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/woopay/class-woopay-session.php [hunk 1]
```diff
@@ -533,9 +533,13 @@ private static function get_woopay_verified_email_address() {
 	 * @return bool True if request is a Store API request, false otherwise.
 	 */
 	private static function is_store_api_request(): bool {
-		$url_parts    = wp_parse_url( esc_url_raw( $_SERVER['REQUEST_URI'] ?? '' ) ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash
-		$request_path = rtrim( $url_parts['path'], '/' );
-		$rest_route   = str_replace( trailingslashit( rest_get_url_prefix() ), '', $request_path );
+		if ( isset( $_REQUEST['rest_route'] ) ) {
+			$rest_route = sanitize_text_field( $_REQUEST['rest_route'] );
```
Comment (bborman22) [inline includes/woopay/class-woopay-session.php:539]: I wasn't sure the best sanitize function to use here?

File: includes/woopay/class-woopay-session.php [hunk 2]
```diff
@@ -533,9 +533,13 @@ private static function get_woopay_verified_email_address() {
 	 * @return bool True if request is a Store API request, false otherwise.
 	 */
 	private static function is_store_api_request(): bool {
-		$url_parts    = wp_parse_url( esc_url_raw( $_SERVER['REQUEST_URI'] ?? '' ) ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash
-		$request_path = rtrim( $url_parts['path'], '/' );
-		$rest_route   = str_replace( trailingslashit( rest_get_url_prefix() ), '', $request_path );
+		if ( isset( $_REQUEST['rest_route'] ) ) {
```
Comment (bborman22) [inline includes/woopay/class-woopay-session.php:538]: This conditional structure favors the `rest_route` parameter over the URI, are there concerns with that approach? For example, could a `rest_route` parameter be present, but we actually wanted to use the URI instead?

**Review by malithsen — APPROVED** [review]
first party auth on + post name permalinks ✅ 
first party auth on + plain permalinks ✅ 
first party auth off + post name permalinks ✅ 
first party auth off + plain permalinks ✅ 

I also did tested with "Month & Name" & "Numeric" permalinks and confirmed the cart loads properly.

Looks good. Thanks @bborman22

File: includes/woopay/class-woopay-session.php (hunk 2 above)
Comment (ricardo) [reply includes/woopay/class-woopay-session.php:538]: It seems like it should be either the URL or the `rest_route` query parameter.

https://github.com/WordPress/WordPress/blob/39ccc3798dc77c98330d3d0f6ac87c1d556e38e1/wp-includes/rest-api/class-wp-rest-request.php#L1036-L1044

**Review by ricardo — APPROVED** [review]
LGTM and tests well.

All test cases below worked with permalinks set to "Plain" or "Post name":
- ✅ Checkout via WooPay button with first party auth on.
- ✅ Checkout via WooPay button with first party auth off.
- ✅ Checkout via email input.

**Pull request merged**